File: backend/app/api/inspections.py
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, Form, status
from sqlalchemy.orm import Session
from typing import List, Optional
from ..database import get_db
from ..models.rbac import User
from ..models.inspection import Inspection, Image
from ..schemas.inspection import InspectionOut, InspectionListOut, InspectionDetailOut
from ..core.dependencies import get_current_user, require_officer_or_admin
from ..services.inspection import create_inspection_record, create_image_record, get_inspection
from ..services.image_storage import save_upload
from ..services.queue import enqueue_inspection_job
from ..services.audit import log_audit_event
from ..config import settings
import uuid

router = APIRouter(prefix="/inspections", tags=["Inspections"])

def validate_image(file: UploadFile):
    # MIME magic byte validation would go here
    # For now, simplistic validation
    if not file.content_type.startswith("image/"):
        raise HTTPException(status_code=400, detail="Only image files are allowed")

@router.post("/", response_model=InspectionOut, status_code=status.HTTP_202_ACCEPTED)
def create_inspection(
    category_id: uuid.UUID = Form(...),
    product_name: str = Form(None),
    location: str = Form(None),
    is_imported: bool = Form(False),
    notes: str = Form(None),
    images: List[UploadFile] = File(...),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_officer_or_admin)
):
    if len(images) == 0 or len(images) > 4:
        raise HTTPException(status_code=400, detail="Must provide between 1 and 4 images")
        
    for img in images:
        validate_image(img)

    # 1. Create inspection record
    inspection = create_inspection_record(
        db=db,
        officer_id=current_user.id,
        category_id=category_id,
        product_name=product_name,
        location=location,
        is_imported=is_imported,
        notes=notes
    )
    
    # 2. Store images
    for idx, img in enumerate(images):
        file_path = save_upload(inspection.id, idx, img, settings.UPLOAD_DIR)
        create_image_record(db, inspection.id, file_path)
        
    # 3. Enqueue Job
    enqueue_inspection_job(inspection.id)
    
    log_audit_event(db, current_user.id, "CREATE_INSPECTION", "inspections", inspection.id)
    
    # Refresh to get DB trigger generated inspection_number
    db.refresh(inspection)
    return inspection

@router.get("/", response_model=InspectionListOut)
def list_inspections(
    skip: int = 0, limit: int = 100,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    query = db.query(Inspection)
    roles = [r.role_name for r in current_user.roles]
    if "ADMIN" not in roles:
        query = query.filter(Inspection.officer_id == current_user.id)
        
    total = query.count()
    inspections = query.order_by(Inspection.created_at.desc()).offset(skip).limit(limit).all()
    
    return {"data": inspections, "total": total}

@router.get("/{inspection_id}", response_model=InspectionDetailOut)
def get_inspection_endpoint(
    inspection_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return get_inspection(db, inspection_id, current_user)

@router.post("/{inspection_id}/retry", status_code=status.HTTP_202_ACCEPTED)
def retry_inspection(
    inspection_id: uuid.UUID,
    images: List[UploadFile] = File(...),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_officer_or_admin)
):
    inspection = get_inspection(db, inspection_id, current_user)
    
    if inspection.status != "PROCESSING_FAILED":
        raise HTTPException(status_code=400, detail="Only failed inspections can be retried by the officer")
        
    if len(images) == 0 or len(images) > 4:
        raise HTTPException(status_code=400, detail="Must provide between 1 and 4 images")
        
    for img in images:
        validate_image(img)
        
    # Count existing images to continue index
    existing_count = db.query(Image).filter(Image.inspection_id == inspection.id).count()
    
    for idx, img in enumerate(images):
        file_path = save_upload(inspection.id, existing_count + idx, img, settings.UPLOAD_DIR)
        create_image_record(db, inspection.id, file_path)
        
    # Reset status
    inspection.status = "PENDING"
    db.commit()
    
    enqueue_inspection_job(inspection.id)
    log_audit_event(db, current_user.id, "RETRY_INSPECTION_WITH_IMAGES", "inspections", inspection.id)
    
    return {"status": "retry_accepted"}

File: frontend/src/api/inspections.ts
import { apiClient } from './client'

export type InspectionStatus =
  | 'PENDING'
  | 'PROCESSING'
  | 'COMPLIANT'
  | 'NON_COMPLIANT'
  | 'REVIEW_REQUIRED'
  | 'PROCESSING_FAILED'

export interface InspectionImage {
  id: string
  file_path: string
  ocr_status: string
}

export interface Finding {
  id: string
  finding_type: string
  status: string
  details?: {
    clause_id?: string
    field_name?: string
    observed_value?: string
    required_value?: string
    reason?: string
    evidence?: {
      image_id?: string
      confidence?: number
      bounding_box?: [number, number, number, number] | null
    }
  }
}

export interface Inspection {
  id: string
  inspection_number: string
  status: InspectionStatus
  product_name?: string
  location?: string
  is_imported: boolean
  notes?: string
  created_at: string
  updated_at: string
  product_category_id: string
  rule_id: string
  officer_id: string
  category_name?: string
  officer_name?: string
  images?: InspectionImage[]
  findings?: Finding[]
  verdict_summary?: string
  error_message?: string
}

export interface InspectionListResponse {
  data: Inspection[]
  total: number
}

export interface CreateInspectionPayload {
  category_id: string
  product_name?: string
  location?: string
  is_imported?: boolean
  notes?: string
  images: File[]
}

export async function createInspection(payload: CreateInspectionPayload): Promise<Inspection> {
  const form = new FormData()
  form.append('category_id', payload.category_id)
  if (payload.product_name) form.append('product_name', payload.product_name)
  if (payload.location) form.append('location', payload.location)
  form.append('is_imported', String(payload.is_imported ?? false))
  if (payload.notes) form.append('notes', payload.notes)
  payload.images.forEach((img) => form.append('images', img))

  const { data } = await apiClient.post<Inspection>('/inspections/', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return data
}

export async function listInspections(skip = 0, limit = 50): Promise<InspectionListResponse> {
  const { data } = await apiClient.get<InspectionListResponse>('/inspections/', {
    params: { skip, limit },
  })
  return data
}

export async function getInspection(id: string): Promise<Inspection> {
  const { data } = await apiClient.get<Inspection>(`/inspections/${id}`)
  return data
}

export async function retryInspection(id: string, images: File[]): Promise<void> {
  const form = new FormData()
  images.forEach((img) => form.append('images', img))
  await apiClient.post(`/inspections/${id}/retry`, form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
}
